feat: implement process definition statistics endpoint for rdbms

tsedekey opened the pull request:
## Description

<!-- Describe the goal and purpose of this PR. -->
This PR adds the RDBMS backend implementation for the Process Definition Statistics API, completing the second data-layer component of the feature. 

Testing for this backend will be covered as part of the upcoming Java client implementation [task](https://github.com/camunda/camunda/issues/36703), including full end-to-end integration tests.

## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [ ] Enable backports when necessary (fex. [for bug fixes](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#backporting-changes) or [for CI changes](https://github.com/camunda/camunda/wiki/CI-&-Automation#when-to-backport-ci-changes)).

## Related issues

closes #36702

tsedekey committed:
feat: add pd statistics version endpoint

fix: correct unintended changes

refactor: remove unused method

refactor: include existing pd ids in filter

test: update test

refactor: address PR comments

refactor: split open api definition

style: fix spotless errors

docs: add license header

test: update test

refactor: address PR comments

tsedekey committed:
feat: add rdbms implementation

fix: merge errors

refactor: correct minor issues

fix: return empty if not authorised

refactor: address co-pilot PR comments

refactor: address CP PR comments

refactor: apply PR comments

fix: add hits check

fix: rearrange code order

style: remove empty space

tsedekey commented on db/rdbms/src/main/java/io/camunda/db/rdbms/read/service/ProcessDefinitionInstanceVersionStatisticsDbReader.java:
```diff
@@ -7,26 +7,68 @@
+
+    final var results = processDefinitionMapper.processInstanceVersionStatistics(dbQuery);
+
+    return buildSearchQueryResult(totalHits, results, dbSort);
```
That’s a really good idea. I will create a follow-up issue for it.

tmetzke replied:
Sounds good 👍

tsedekey committed:
refactor: adding abstract method to execute paged query

tmetzke commented on db/rdbms/src/main/java/io/camunda/db/rdbms/read/service/AbstractEntityReader.java:
```diff
@@ -173,4 +173,27 @@ protected boolean shouldReturnEmptyResult(final ResourceAccessChecks resourceAcc
+   * @return a SearchQueryResult containing the results and total count
+   */
+  protected SearchQueryResult<T> executePagedQuery(
+      final java.util.function.Supplier<Long> countSupplier,
```
🙃 Any specific reason not to add the `Supplier` to the import section? 🙂

tmetzke commented on db/rdbms/src/main/java/io/camunda/db/rdbms/read/service/ProcessDefinitionInstanceVersionStatisticsDbReader.java:
```diff
@@ -7,26 +7,65 @@
+    final var dbSort =
+        convertSort(
+            query.sort(),
+            ProcessDefinitionInstanceVersionStatisticsSearchColumn.PROCESS_DEFINITION_ID);
```
💭 This is the discriminator column in case of otherwise identical values. Isn't the `VERSION` the discriminator here? 🤔

tmetzke commented on db/rdbms/src/main/java/io/camunda/db/rdbms/sql/ProcessDefinitionMapper.java:
```diff
@@ -24,4 +28,15 @@ public interface ProcessDefinitionMapper {
+      ProcessDefinitionInstanceVersionStatisticsDbQuery filter);
+
+  List<ProcessDefinitionInstanceVersionStatisticsEntity> processInstanceVersionStatistics(
+      ProcessDefinitionInstanceVersionStatisticsDbQuery filter);
```
🔧 We're passing `query`, not `filter`, I guess.

tmetzke commented on db/rdbms/src/main/resources/mapper/ProcessDefinitionMapper.xml:
```diff
@@ -320,4 +320,174 @@
+        )
+      </if>
+    </where>
+    GROUP BY pd.PROCESS_DEFINITION_ID
```
❌ We must also group by tenant ID. You can have a look at how `isLatestVersion` considers this as well.

tmetzke commented on db/rdbms/src/main/resources/mapper/ProcessDefinitionMapper.xml:
```diff
@@ -320,4 +320,174 @@
+        )
+      </if>
+    </where>
+    GROUP BY pd.PROCESS_DEFINITION_ID
```
❌ We must also group by tenant ID.

tmetzke commented on db/rdbms/src/main/resources/mapper/ProcessDefinitionMapper.xml:
```diff
@@ -320,4 +320,174 @@
+          #{tid}
+        </foreach>
+      </if>
+      <include refid="io.camunda.db.rdbms.sql.ProcessInstanceMapper.searchFilter"/>
```
🔧 Same as for the parent stats, we only need ACTIVE instances, but we don't need the whole filter.

tmetzke reviewed (commented):
👍 Really good direction already 🚀 

💭 Overall, maybe we can ignore a generic process instance filter in the search layer as well for now and hardwire the conditions in the ES/OS/RDBMS aggregation transformers? That would allow creating more focused queries/mappings, I believe:
- For the parent statistics, we group by definition ID and tenant ID. We don't consider tenant ID here yet, I think, but two process definitions can have the same ID if they are on different tenants, so we need to group by both.
- For both queries, we need to hardwire to ACTIVE process instances (by the way, I think we don't do this for the parent query right now in the `ProcessDefinitionServices`). But that is all we need. We don't need to hand this in as a process instance filter.
- For the version statistics, we also need to hardwire to the `processDefinitionId` and `tenantId` of the parent (same reason as for grouping the parents by tenant ID). This way, we get the correct number of process instances for each version.
- This should simplify our RDBMS mappings. For example, for the version statistics, we only need to group by version, then, if I'm not mistaken? 🤔 

💭 Having QA acceptance tests in this PR as well could be super helpful to validate the queries and approach. Maybe we should tackle the client implementation before we merge this. We can also follow up in the QA PR, of course, but it feels like we can't easily validate our approach here right now.

tsedekey committed:
refactor: applied various changes, see description

- process definition instance statistics:
  - Remove filter from REST API
  - ACTVIE state filter is applied the service layer for ES/OS and SQL
mapper layer for RDBMS.
  - Simplfied SQL and added grouping by tenants

tsedekey committed:
refactor: address PR comments

tsedekey committed:
refactor: added grouping by tenant for doc data

tsedekey committed:
refactor: add tenantId filter to version endpoint

tsedekey committed:
refactor: add tenantId grouping and filter to rdbms

tsedekey committed:
fix: fix rebase error

tsedekey committed:
style: fix spotless errors

tsedekey committed:
test: update tests

tsedekey committed:
test: update ut

tsedekey committed:
refactor: fix sql query and test

tsedekey committed:
test: add test to validate field and script

tsedekey committed:
test: update failed test

tsedekey committed:
test: update OS test

tmetzke commented on db/rdbms/src/main/resources/mapper/ProcessDefinitionMapper.xml:
```diff
@@ -320,4 +320,147 @@
+    resultType="io.camunda.search.entities.ProcessDefinitionInstanceStatisticsEntity">
+    SELECT pi.PROCESS_DEFINITION_ID,
+    pi.TENANT_ID,
+    MAX(pd.NAME) AS LATEST_PROCESS_DEFINITION_NAME,
```
💭 Interesting. Ideally, this is the name of the latest version, not the maximum of the names, right? 🤔

tmetzke commented on db/rdbms/src/main/resources/mapper/ProcessDefinitionMapper.xml:
```diff
@@ -320,4 +320,147 @@
+    pi.PROCESS_DEFINITION_KEY
+    FROM
+    ${prefix}PROCESS_INSTANCE pi
+    LEFT JOIN
```
💭 Using a LEFT JOIN will leave us with process instance entries that cannot be matched to process definitions by key, right? Does that make sense here? Would a simple JOIN be more appropriate? 🤔

tmetzke reviewed (commented):
👍 Thanks for maneuvering this! This is going in the right direction 🚀 
💭 I left a couple of thoughts for the RDBMS mapper. Nothing huge, though.
🧪 Ultimately, we'll only really know once we have data-driven acceptance tests. Let's ensure we're solid on those to identify gaps we may have here.

tmetzke commented on db/rdbms/src/main/resources/mapper/ProcessDefinitionMapper.xml:
```diff
@@ -320,4 +320,147 @@
+      WHERE pi2.PROCESS_DEFINITION_ID = pi.PROCESS_DEFINITION_ID
+      ORDER BY pi2.VERSION DESC
+      LIMIT 1
+    )
```
I see. OK, let's go for the correlated subquery for now then. In the end, it's an implementation detail we can follow-up on with a load test, for example.

tsedekey committed:
refactor: re-use sql fragments

tsedekey committed:
fix: update authorization logic after merge

tmetzke reviewed (approved):
👍 Let's go 🚀 
💭 The acceptance tests in the next PR will reveal any limitations that we might have overlooked. Let's make sure those are solid!

tmetzke commented:
Yeah, I wanted to avoid this one becoming stale and running into merge conflicts.
Other teams will bring PR that deal with similar code sections. Maybe you can quickly skim https://github.com/camunda/camunda/pull/39986 as well to see if anything you worked on here should be streamlined there. Even if that one is merged before you get to it, leaving comments there so Jonathan can follow up with another PR helps as well 👍
